auction-server/src/opportunity/service/remove_invalid_or_expired_opportunities.rs

```diff
@@ -63,3 +63,209 @@ impl Service {
+            }),
+        }
+    }
+
+    async fn push_test_opportunity(service: &Service, test_opportunity: OpportunitySvm) {
+        service
+            .repo
+            .in_memory_store
+            .opportunities
+            .write()
+            .await
+            .entry(test_opportunity.get_key())
+            .or_insert_with(Vec::new)
+            .push(test_opportunity);
+    }
+
+    #[tokio::test]
+    async fn test_dont_remove_valid_opportunities() {
+        let rpc_client = RpcClientSvmTester::new();
+        let mut mock_db = MockDatabase::default();
+
+        mock_db.expect_add_opportunity().returning(|_| Ok(()));
+
+        let (service, ws_receiver) =
+            Service::new_with_mocks_svm(DEFAULT_CHAIN_ID.to_string(), mock_db, &rpc_client);
+
+        // recent opportunity
+        let test_time = OffsetDateTime::now_utc() - Duration::minutes(1);
+        let test_opportunity = make_test_opportunity(1, test_time, test_time);
+
+        push_test_opportunity(&service, test_opportunity.clone()).await;
+
+        service.remove_invalid_or_expired_opportunities().await;
+
+        // check opportunities
+        let all_opportunities_by_test_key = service
+            .repo
+            .get_in_memory_opportunities()
+            .await
+            .remove(&test_opportunity.get_key())
+            .expect("opportunity should exist");
+
+        assert_eq!(all_opportunities_by_test_key, vec![test_opportunity]);
+        assert!(ws_receiver.is_empty());
+    }
+
+    #[tokio::test]
+    async fn test_remove_one_expired_opportunity() {
+        let rpc_client = RpcClientSvmTester::new();
+        let mut mock_db = MockDatabase::default();
+
+        mock_db.expect_add_opportunity().returning(|_| Ok(()));
+        mock_db.expect_remove_opportunity().returning(|_, _| Ok(()));
+
+        let (service, ws_receiver) =
+            Service::new_with_mocks_svm(DEFAULT_CHAIN_ID.to_string(), mock_db, &rpc_client);
+
+        // recent opportunity
+        let time_alive = OffsetDateTime::now_utc() - Duration::minutes(1);
+        let time_expired = OffsetDateTime::now_utc() - Duration::minutes(5);
+        let live_opportunity = make_test_opportunity(1, time_alive, time_alive);
+        let expired_opportunity = make_test_opportunity(2, time_expired, time_expired);
+
+        push_test_opportunity(&service, live_opportunity.clone()).await;
+        push_test_opportunity(&service, expired_opportunity.clone()).await;
+
+        service.remove_invalid_or_expired_opportunities().await;
+
+        // check opportunities
+        let all_opportunities_by_test_key = service
+            .repo
+            .get_in_memory_opportunities()
+            .await
+            .remove(&live_opportunity.get_key())
+            .expect("opportunity should exist");
+
+        assert_eq!(all_opportunities_by_test_key, vec![live_opportunity]);
+        assert!(ws_receiver.is_empty());
```

Comment:
why is this empty, didn't we remove `expired_opportunity`

Reply:
we removed one, but the point is that we didnt remove all of them under the permission key, this is specifically what the code is checking for, not sure if its sound logic but it works